File: src/main/java/ccpl/lib/Bundle.java
/*
 * This file is part of the Cohen Ray Number Line.
 *
 * Latesco is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Latesco is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Latesco.  If not, see <http://www.gnu.org/licenses/>.
 */

package ccpl.lib;

import static java.lang.Boolean.parseBoolean;
import static java.lang.Integer.parseInt;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class Bundle {
  private Map<String, Object> values = new HashMap<>();

  public void add(String key, Object value) {
    values.put(key, value);
  }

  private Object get(String key) {
    return values.get(key);
  }

  public String getAsString(String key) {
    return get(key).toString();
  }

  public Integer getAsInt(String key) {
    return parseInt(getAsString(key));
  }

  public Boolean getAsBoolean(String key) {
    return parseBoolean(getAsString(key));
  }

  /**
   * Merges two Bundle together, destructively.
   *
   * <p>This function favors values found in the first Bundle.
   *
   * @param secondBundle Second bundle
   * @return New merged Bundle.
   */
  public Bundle merge(Bundle secondBundle) {
    Bundle newBundle = new Bundle();

    secondBundle.values.forEach(newBundle::add);
    values.forEach(newBundle::add);

    return newBundle;
  }

  public boolean contains(String key) {
    return values.containsKey(key);
  }

  public int getSize() {
    return values.size();
  }

  @Override
  public String toString() {
    List<String> vals =
        values
            .entrySet()
            .stream()
            .map(it -> it.getKey() + ": " + it.getValue())
            .collect(Collectors.toCollection(ArrayList::new));

    return String.join("\n", vals);
  }
}

File: src/main/java/ccpl/lib/Bundle.kt
package ccpl.lib


class Bundle {
  private val values: MutableMap<String, Any> = hashMapOf()

  val size
    get() = values.size

  fun add(key: String, value: Any) {
    values.put(key, value)
  }

  private fun get(key: String): Any? {
    return if (!values.containsKey(key)) {
      throw Exception("Bundle does not contain key: " + key)
    } else {
      values[key]
    }
  }

  fun getAsString(key: String) = get(key) as String

  fun getAsInt(key: String) = if (get(key) is String) Integer.parseInt(getAsString(key))
  else get(key) as Int

  fun getAsBoolean(key: String): Boolean = if (get(key) is String) getAsString(key).toBoolean()
  else get(key) as Boolean

  fun merge(secondBundle: Bundle) : Bundle {
    val newBundle = Bundle()

    secondBundle.values.forEach { newBundle.add(it.key, it.value) }
    values.forEach { newBundle.add(it.key, it.value) }

    return newBundle
  }

  override fun toString(): String = values.map { it.key + ": " + it.value }.joinToString("\n")

  fun contains(key: String) = values.contains(key)
}